godot-macros/src/util/list_parser.rs

```diff
@@ -0,0 +1,206 @@
+
+            return bail!(group.span_open(), "expected `{expected}`, got `{got}`");
+        }
+
+        let trees: Vec<TokenTree> = group.stream().into_iter().collect();
+
+        let raw_lists = trees
+            .split_inclusive(|tree| is_punct(tree, ','))
+            .collect::<Vec<_>>();
+
+        let list_len = raw_lists.len();
+        let mut lists = Vec::new();
+
+        for (i, list) in raw_lists.into_iter().enumerate() {
+            let is_last = i == list_len - 1;
+
+            // every list except the last one must contain at least a comma so this means we have a trailing
+            // comma.
+            if list.is_empty() {
+                break;
+            }
+
+            // does list only contain `,`?
+            if !is_last && list.len() == 1 {
+                let list_stream = list.iter().cloned().collect::<TokenStream>();
+                return bail!(list_stream, "expected expression, found `,`");
+            }
+
+            let end = if is_last { list.len() } else { list.len() - 1 };
+
+            lists.push(KvValue::new((&list[..end]).into()));
+        }
+
+        Ok(Self {
+            lists: lists.into(),
+            span_close: group.span_close(),
+        })
+    }
+
+    /// Get the next element from the list, starting at the front.
+    fn pop_next(&mut self) -> Option<KvValue> {
+        self.lists.pop_front()
+    }
+
+    fn peek(&self) -> Option<&KvValue> {
+        self.lists.front()
+    }
+
+    /// Take the next element of the list, ensuring it is an expression.
+    pub fn next_expr(&mut self) -> ParseResult<TokenStream> {
+        let Some(kv) = self.pop_next() else {
+            return bail!(self.span_close, "expected expression");
+        };
+
+        kv.expr()
+    }
+
+    /// Take the next element of the list, if it is an identifier.
+    ///
+    /// Returns `Ok(None)` if there are no more elements left.
+    pub fn next_ident(&mut self) -> ParseResult<Option<Ident>> {
+        let Some(kv) = self.pop_next() else { return Ok(None) };
+
+        Ok(Some(kv.ident()?))
+    }
+
+    /// Take the next element of the list, if it is an identifier.
+    ///
+    /// Returns `Ok(None)` if there are no more elements left or the next element isn't an identifier.
+    pub fn try_next_ident(&mut self) -> ParseResult<Option<Ident>> {
+        let Some(kv) = self.peek() else { return Ok(None) };
+
+        let id = kv.as_ident()?;
+
+        _ = self.pop_next();
+
+        Ok(Some(id))
+    }
+
+    /// Take the next element of the list, ensuring it is one of the given identifiers.
+    ///
+    /// Returns `Ok(None)` if there are no more elements left.
+    pub fn next_any_ident(&mut self, ids: &[&str]) -> ParseResult<Option<Ident>> {
+        let Some(next_id) = self.try_next_ident()? else {
+            return Ok(None)
+        };
+
+        for id in ids {
+            if ident(id) == next_id {
+                return Ok(Some(next_id));
+            }
```

Comment:
I think we could just do `if next_id == id`, not needing to create an extra `ident(...)`.